unlock-protocol.com/blog/emergency-upgrade-1-3.md

```diff
@@ -0,0 +1,116 @@
+---
+title: Emergency Upgrade
+subTitle: Version 1-3 is now available
+authorName: Nick Mancuso
+publishDate: January 17, 2020
+description: Explaining what went wrong and why we released a new version so quickly after our last upgrade.
+image: /static/images/blog/unlocking-smart-contracts/code.jpeg
+---
+
+We recently noticed a vulnerability which if abused could have had bad consequences – but we caught it early and fixed it promptly. This post aims to explain the errors we made with our most recent smart-contract update.
+
+There were problems with this release which if targeted by an attacker had the potential to do one of the following:
+ - Steal money made from previous key sales.
+ - Break every lock (and key) using the new version - blocking new sales, preventing key owners from being recognized and preventing the lock owner from accessing any funds previously made.
+
+We fixed the second issue already, so no worries there.  But funds are at risk if your lock is impacted.  For that we need the lock owners themselves to upgrade ASAP.
+
+This only impacts locks created in the past couple weeks. If you recently created a lock, please go to the Unlock dashboard to check if you are impacted.  The dashboard will flag any impacted lock, asking you to upgrade.
+
+So what happened?  Let’s start with the issue we were able to fix right away and then we’ll get to the juicy stuff.
+
+## Break every lock using the new version
+
+> This issue is about our deployment process.  We simply needed to take action before an attacker.  We will avoid this race condition in the future.
+
+We recently changed the locks deployed to use the minimal proxy standard, [EIP-1167](https://eips.ethereum.org/EIPS/eip-1167).  This significantly reduces the cost to create new locks and leads to less blockchain bloat.  When using proxies any data set in the constructor is lost so instead we call a single-use function called `initialize`.
+
+During normal operation a user calls `createLock` on our main Unlock contract.  That function will deploy the lock and call `initialize` in the same transaction.  It’s responsible for recording data such as the lock owner and the key price.
+
+Proxies work by reading the implementation logic from a template.  The template is nothing more than a contract containing the logic / source code for the lock proxies to leverage.
+
+We forgot to `initialize` the template!
+
+If someone figured this out before we did they could have called `initialize` on the template to make themselves the owner and then make an owner-only call to `selfdestruct` the contract.  
```

Comment:
I wonder if we should mention why we chose to keep `selfdestruct` in the first place? Minor thing, but it was a conscious decision with the goal of helping minimize bloat on the blockchain.

Reply:
openzeppelin even has a warning about this here: https://docs.openzeppelin.com/upgrades/2.6/writing-upgradeable#potentially-unsafe-operations

Reply:
I think it's worth adding this link to the blog post!

Reply:
Added both, thanks